### PR TITLE
rfc: batch normalization: support only scale memory argument

### PR DESCRIPTION
**Pull request by dzarukin**
This RFC is to discuss options how to improve usability for a single scale memory argument when no shifts are used.

[Link to a rendered document](https://github.com/dzarukin/oneDNN/blob/dzarukin/rfcs/batch_norm_shift_scale_support_v2/rfcs/20210223-batch-norm-scale-only/README.md).

### CODE REVIEW

**Review by igorsafo — APPROVED** [review]

**Review by sbogusev — APPROVED** [review]
I vote for pursuing Option 1 since it is very explicit and clear for the user. We can revisit using Option 2 in case we identify some more performance optimization opportunity.

File: rfcs/20210223-batch-norm-scale-only/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+/// A special mnemonic for shift argument of normalization primitives.
+#define DNNL_ARG_SHIFT 52 // or alternatively DNNL_ARG_WEIGHTS_2 (35)
+~~~
+
+For compatibility reasons, oneDNN would still support both versions of flags
+but internally would rely only on new ones. It means that using
+`dnnl_use_scaleshift` is same as using `dnnl_use_scale | dnnl_use_shift`.
+
+As for memory arguments, pointers will be taken from argument map based on flag
+value specified.
```
[the remark this answers is not in the document]
Comment (dzarukin) [reply rfcs/20210223-batch-norm-scale-only/README.md:65]: I don't feel using `dnnl_use_scaleshift` with `DNNL_ARG_SCALE` + `DNNL_ARG_SHIFT` makes sense, thus, will be restricted.
In second case the execution will abort due to lack of argument since `DNNL_ARG_SCALE` wasn't specified.

This super combo is only needed to keep API intact for those who's already using it, but by design double flag for doubled input, separate flags for separated inputs.

Would you agree with such approach?

[the remark this answers is not in the document]
Comment (mgouicem) [reply rfcs/20210223-batch-norm-scale-only/README.md:65]: Ok that makes sense. Just wasn't completely clear for me from the text. Thanks!

**Review by mgouicem — APPROVED** [review]

**Review by mgouicem — APPROVED** [review]
Ok with option 1 as well.

File: rfcs/20210223-batch-norm-scale-only/README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+/// A special mnemonic for scale argument of normalization primitives.
+#define DNNL_ARG_SCALE 51 // or alternatively DNNL_ARG_WEIGHTS_1 (34)
+/// A special mnemonic for shift argument of normalization primitives.
+#define DNNL_ARG_SHIFT 52 // or alternatively DNNL_ARG_WEIGHTS_2 (35)
+~~~
+
+For compatibility reasons, oneDNN would still support both versions of flags
+but internally would rely only on new ones. It means that using
+`dnnl_use_scaleshift` is same as using `dnnl_use_scale | dnnl_use_shift`.
```
Comment (guizili0) [inline rfcs/20210223-batch-norm-scale-only/README.md:62]: @mgouicem so when using scale and shift as two separate memory, I need set  `dnnl_use_scale | dnnl_use_shift` and give two memory object in the execution args, is it right?

Comment (guizili0) [reply rfcs/20210223-batch-norm-scale-only/README.md:62]: I also vote option one.

Comment (dzarukin) [reply rfcs/20210223-batch-norm-scale-only/README.md:62]: > @mgouicem so when using scale and shift as two separate memory, I need set `dnnl_use_scale | dnnl_use_shift` and give two memory object in the execution args, is it right?

Hi @guizili0, thanks for the question. Yes, this is how it will be designed.

Comment (guizili0) [reply rfcs/20210223-batch-norm-scale-only/README.md:62]: got it, thank you!
